diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,410 @@
+        self,
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: Y,
+        y1_candidate: Y,
+        args: Args,
+        y_error: Optional[Y],
+        error_order: RealScalarLike,
+        controller_state: _JumpStepState[_ControllerState],
+    ) -> tuple[
+        BoolScalarLike,
+        RealScalarLike,
+        RealScalarLike,
+        BoolScalarLike,
+        _JumpStepState[_ControllerState],
+        RESULTS,
+    ]:
+        (
+            made_jump,
+            prev_dt,
+            i_step,
+            i_jump,
+            i_rjct,
+            rjct_buff,
+            step_ts,
+            jump_ts,
+            inner_state,
+        ) = controller_state.get()
+
+        # Let the controller do its thing
+        (
+            keep_step,
+            next_t0,
+            next_t1,
+            _,
+            inner_state,
+            result,
+        ) = self.controller.adapt_step_size(
+            t0, t1, y0, y1_candidate, args, y_error, error_order, inner_state
+        )
+
+        # This is just a logging utility for testing purposes
+        if self.callback_on_reject is not None:
+            jax.debug.callback(self.callback_on_reject, keep_step, t1)
+
+        # Check whether we stepped over an element of step_ts or jump_ts or rjct_buff
+        # This is all still bookkeeping for the PREVIOUS STEP.
+        if step_ts is not None:
+            # If we stepped to `t1 == step_ts[i_step]` and kept the step, then we
+            # increment i_step and move on to the next t in step_ts.
+            step_inc_cond = keep_step & (t1 == _get_t(i_step, step_ts))
+            i_step = jnp.where(step_inc_cond, i_step + 1, i_step)
+
+        if jump_ts is not None:
+            next_jump_t = _get_t(i_jump, jump_ts)
+            jump_inc_cond = keep_step & (t1 >= eqxi.prevbefore(next_jump_t))
+            i_jump = jnp.where(jump_inc_cond, i_jump + 1, i_jump)
+
+        if self.rejected_step_buffer_len > 0:
+            assert rjct_buff is not None
+            # If the step ended at t1==rjct_buff[i_rjct], then we have successfully
+            # stepped to this time and we increment i_rjct.
+            # We increment i_rjct even if the step was rejected, because we will
+            # re-add the rejected time to the buffer immediately.
+            rjct_inc_cond = t1 == _get_t(i_rjct, rjct_buff)
+            i_rjct = jnp.where(rjct_inc_cond, i_rjct + 1, i_rjct)
+
+            # If the step was rejected, then we need to store the rejected time in the
+            # rejected buffer and decrement the rejected index.
+            i_rjct = jnp.where(keep_step, i_rjct, i_rjct - 1)
+            i_rjct = eqx.error_if(
+                i_rjct,
+                i_rjct < 0,
+                "Maximum number of rejected steps reached. "
+                "Consider increasing JumpStepWrapper.rejected_step_buffer_len.",
+            )
+            rjct_buff = jnp.where(keep_step, rjct_buff, rjct_buff.at[i_rjct].set(t1))
```

Comment:
Note that this is a very expensive way to describe this operation! You're copying the whole buffer. XLA will *sometimes* optimize this out -- because I added that optimization to it! -- but not always.

Better is to do `rjct_buff.at[i_rjct].set(jnp.where(keep_step, rjct_buff[i_rjct], t1))`

Reply:
Other than that, I think we may need to extend the API here slightly -- we should be able to mark state like this as being a buffer for the purposes of:

https://github.com/patrick-kidger/diffrax/blob/0679807bf7bda0de3884a78cf20ce2dd4d8ade4c/diffrax/_integrate.py#L621

which is needed to avoid spurious copies during backpropagation.

(You can see that both of these comments are basically us having to work around limitations of the XLA compiler.)

Reply:
The first one makes sense, I should have seen this.

I don't really know what you want me to do in your second comment. And frankly `diffeqsolve` is something I haven't even started digesting yet. Are you telling me `rejected_buffer` should be one of the `outer_buffers`, meaning that I should make it an instance of `SaveState` or sth like that? I would apprecaite a bit more guidance.

Also damn how you managed to write all this code is beyond me. Even trying to begin understanding it seems a lot!

Reply:
Haha, you're too kind!

As for my second comment -- I think I've realised that I was wrong. Let me explain. Our backpropagation involves saving copies of our state in checkpoints. Let's suppose we set `RecursiveCheckpointAdjoint(checkpoints=max_steps)`, so that's `O(max_steps)` memory right? Well, not quite: our updating buffer here is potentially of length `max_steps` (as per the debate above), and we're saving a copy of it in every checkpont, so we'd actually be using `O(max_steps^2)` memory!

The simple solution to this will just be to set the size of this buffer to e.g. 100 by default, and just allow those copies to be made. And given the behaviour you have here -- in which you potentially overwrite values -- then that is actually what's necessary as well.

As for the complicated solution that I was wrong about: let's consider the case of `SaveAt(steps=True)`. This also involves a buffer of length `max_steps`, that we save into as we go along. Fortunately, this one has a useful extra property, which is that we never overwrite a value. That means we don't actually need to copy our buffer for every checkpoint! We can use a single buffer that is shared across all checkpoints, getting gradually filled in. To support this case then we actually have a special argument `eqxi.while_loop(..., buffers=...)`, to declare which elements of our loop state have this behaviour. Unfortunately that's not the case here because we do overwrite the values. (And side-note the presence of this `buffers` parameter is the reason I've not made this public API in Equinox, because the buffer-ness is completely unchecked and it's very easy to shoot yourself in the foot.)

Reply:
Okay, I see. Thanks for the in-depth explanation! So let's see if I understand this correctly:

- I should set the default value to be 100 and keep the type as `Optional[int]`.
- If this was not getting rewritten, then I should make it register as a buffer in the outer while loop of diffeqsolve. But, because it does get rewritten, I should not do that(??). Still, I am curious, if I did want to register it as a buffer, how would I accomplish that? Is it indeed by making it an instance of `SaveState`, or is it something else entirely?